src/library/scala/reflect/ClassTag.scala

```diff
@@ -111,29 +111,36 @@ object ClassTag {
   val Nothing : ClassTag[scala.Nothing]    = Manifest.Nothing
   val Null    : ClassTag[scala.Null]       = Manifest.Null
 
+  private[this] val cache = new ClassValue[ClassTag[_]] {
+    override def computeValue(runtimeClass: jClass[_]): ClassTag[_] = new GenericClassTag[AnyRef](runtimeClass)
+  }
+
   @SerialVersionUID(1L)
   private class GenericClassTag[T](val runtimeClass: jClass[_]) extends ClassTag[T] {
     override def newArray(len: Int): Array[T] = {
       java.lang.reflect.Array.newInstance(runtimeClass, len).asInstanceOf[Array[T]]
     }
   }
 
-  def apply[T](runtimeClass1: jClass[_]): ClassTag[T] =
-    runtimeClass1 match {
-      case java.lang.Byte.TYPE      => ClassTag.Byte.asInstanceOf[ClassTag[T]]
-      case java.lang.Short.TYPE     => ClassTag.Short.asInstanceOf[ClassTag[T]]
-      case java.lang.Character.TYPE => ClassTag.Char.asInstanceOf[ClassTag[T]]
-      case java.lang.Integer.TYPE   => ClassTag.Int.asInstanceOf[ClassTag[T]]
-      case java.lang.Long.TYPE      => ClassTag.Long.asInstanceOf[ClassTag[T]]
-      case java.lang.Float.TYPE     => ClassTag.Float.asInstanceOf[ClassTag[T]]
-      case java.lang.Double.TYPE    => ClassTag.Double.asInstanceOf[ClassTag[T]]
-      case java.lang.Boolean.TYPE   => ClassTag.Boolean.asInstanceOf[ClassTag[T]]
-      case java.lang.Void.TYPE      => ClassTag.Unit.asInstanceOf[ClassTag[T]]
-      case ObjectTYPE               => ClassTag.Object.asInstanceOf[ClassTag[T]]
-      case NothingTYPE              => ClassTag.Nothing.asInstanceOf[ClassTag[T]]
-      case NullTYPE                 => ClassTag.Null.asInstanceOf[ClassTag[T]]
-      case _                        => new GenericClassTag[T](runtimeClass1)
-    }
+  def apply[T](runtimeClass1: jClass[_]): ClassTag[T] = runtimeClass1 match {
+    case x if x.isPrimitive =>  primitiveClassTag(runtimeClass1)
```

Comment:
is it better to have this logic in the `computeValue`
depends on the frequency of lookup etc

Ideally we could add the special values to the cache on its construction, but I can see a hook to allow that in `ClassValue`

Reply:
That sounds more regular, I've added a commit.

It slightly improves the lookup speed of refererence classtags.

```
[info] ClassTagBenchmark.lookupClassTag  avgt   20  33.653 ± 0.644  ns/op
```